interfaces/seccomp/backend_test.go

```diff
@@ -392,3 +420,103 @@ func (s *backendSuite) TestSandboxFeatures(c *C) {
+func (s *backendSuite) TestRequiresSocketcallByNotNeededArch(c *C) {
+	testArchs := []string{"amd64", "armhf", "arm64", "powerpc", "ppc64el", "unknownDefault"}
+	for _, arch := range testArchs {
+		restore := seccomp.MockUbuntuKernelArchitecture(func() string { return arch })
+		defer restore()
+		c.Assert(seccomp.RequiresSocketcall(""), Equals, false)
+	}
+}
+
+func (s *backendSuite) TestRequiresSocketcallForceByArch(c *C) {
+	testArchs := []string{"sparc", "sparc64"}
+	for _, arch := range testArchs {
+		restore := seccomp.MockUbuntuKernelArchitecture(func() string { return arch })
+		defer restore()
+		c.Assert(seccomp.RequiresSocketcall(""), Equals, true)
+	}
+}
+
+func (s *backendSuite) TestRequiresSocketcallForcedViaUbuntuRelease(c *C) {
+	testDistros := []string{"ubuntu", "other"}
+	testArchs := []string{"i386", "s390x", "other"}
+	testReleases := []string{"14.04", "other"}
+	for _, distro := range testDistros {
+		restore := seccomp.MockReleaseInfoId(distro)
+		defer restore()
+		for _, arch := range testArchs {
+			restore := seccomp.MockUbuntuKernelArchitecture(func() string { return arch })
+			defer restore()
+			for _, release := range testReleases {
+				restore = seccomp.MockReleaseInfoVersionId(release)
+				defer restore()
+
+				expected := true
+				if distro == "other" || arch == "other" || release == "other" {
+					expected = false
+				}
+				// specify "core18" here so as not to influence
+				// the release check.
+				c.Assert(seccomp.RequiresSocketcall("core18"), Equals, expected)
+			}
+		}
+	}
+}
+
+func (s *backendSuite) TestRequiresSocketcallForcedViaKernelVersion(c *C) {
+	restore := seccomp.MockReleaseInfoId("other")
+	defer restore()
+
+	testArchs := []string{"i386", "s390x", "other"}
+	testVersions := []string{"4.2", "4.3", "4.4"}
+	for _, arch := range testArchs {
+		restore = seccomp.MockUbuntuKernelArchitecture(func() string { return arch })
+		defer restore()
+		for _, version := range testVersions {
+			restore = seccomp.MockKernelVersion(func() string { return version })
+			defer restore()
+
+			expected := false
+			if (arch == "i386" || arch == "s390x") && version == "4.2" {
+				expected = true
+			}
+			// specify "core18" here so as not to influence the
+			// kernel check.
+			c.Assert(seccomp.RequiresSocketcall("core18"), Equals, expected)
+		}
+	}
+}
+
+func (s *backendSuite) TestRequiresSocketcallForcedViaBaseSnap(c *C) {
+	// Mock up as non-Ubuntu, i386 with new enough kernel so the base snap
+	// check is reached
+	restore := seccomp.MockReleaseInfoId("other")
+	defer restore()
+	restore = seccomp.MockUbuntuKernelArchitecture(func() string { return "i386" })
+	defer restore()
+	restore = seccomp.MockKernelVersion(func() string { return "4.3" })
+	defer restore()
+
+	testBases := []string{"", "core", "core16"}
+	for _, baseSnap := range testBases {
+		c.Assert(seccomp.RequiresSocketcall(baseSnap), Equals, true)
+	}
+}
+
+func (s *backendSuite) TestRequiresSocketcallNotForcedViaBaseSnap(c *C) {
```

Comment:
I was thinking about this a bit and experimenting how to make this into a single table driven tests with wildcards. Below is what I came up with. However I'm not super happy with the result, it does feel a bit too magic and non-obvious as a test (I like the table though but the implementation is a bit hard to read). So probably not worth adding.
```
func (s *backendSuite) TestRequiresSocketcall(c *C) {
	tests := []struct {
		arch    string
		distro  string
		release string
		kver    string
		base    string
	}{
		// for i386/s390x on ubuntu 14.04 we always need socketcall
		{"i386", "ubuntu", "14.04", "*", "*"},
		{"s390x", "ubuntu", "14.04", "*", "*"},
		// for i386/s390x with kernel < 4.2 we always need socketcall
		{"i386", "*", "*", "4.2", "*"},
		{"s390x", "*", "*", "4.2", "*"},
		// for i386/s390x with bases "", "core", "core16" we always need socketcall
		{"i386", "*", "*", "*", ""},
		{"s390x", "*", "*", "*", ""},
		{"i386", "*", "*", "*", "core"},
		{"s390x", "*", "*", "*", "core"},
		{"i386", "*", "*", "*", "core16"},
		{"s390x", "*", "*", "*", "core16"},
		// for sparc/sparc64 we always need socketcall
		{"sparc", "*", "*", "*", "*"},
		{"sparc64", "*", "*", "*", "*"},
	}

	// everything else is negative
	requiresSocketcall := func(arch, distro, release, kver, base string) bool {
		for _, t := range tests {
			if (t.arch == arch || t.arch == "*") && (t.distro == distro || t.distro == "*") && (t.release == release || t.release == "*") && (t.kver == kver || t.kver == "*") && (t.base == base || t.base == "*") {
				return true
			}
		}
		return false
	}

	reqs := make([]string, 0, 200)
	for _, arch := range []string{"i386", "s390x", "amd64", "armhf", "arm64", "powerpc", "ppc64el", "sparc", "sparc64", "arch-other"} {
		for _, distro := range []string{"ubuntu", "distro-other"} {
			for _, release := range []string{"14.04", "rel-other"} {
				for _, kver := range []string{"4.2", "4.3", "kver-other"} {
					for _, base := range []string{"", "core", "core16", "base-other"} {
						restore := seccomp.MockUbuntuKernelArchitecture(func() string { return arch })
						defer restore()
						restore = seccomp.MockReleaseInfoId(distro)
						defer restore()
						restore = seccomp.MockReleaseInfoVersionId(release)
						defer restore()
						restore = seccomp.MockKernelVersion(func() string { return kver })
						defer restore()

						expected := requiresSocketcall(arch, distro, release, kver, base)
						c.Check(seccomp.RequiresSocketcall(base), Equals, expected, Commentf("incorrect result for %v", []string{arch, distro, release, kver, base}))
						if expected {
							reqs = append(reqs, fmt.Sprintf("%s/%s/%s/%s/%s", arch, distro, release, kver, base))
						}
					}
				}
			}
		}
	}
	// sanity check
	c.Check(reqs, testutil.Contains, "i386/ubuntu/14.04/kver-other/base-other")
	c.Check(reqs, testutil.Contains, "s390x/ubuntu/14.04/kver-other/base-other")
	c.Check(reqs, testutil.Contains, "i386/ubuntu/rel-other/4.2/base-other")
	c.Check(reqs, testutil.Contains, "s390x/ubuntu/rel-other/4.2/base-other")
	c.Check(reqs, testutil.Contains, "sparc/distro-other/rel-other/kver-other/base-other")
	c.Check(reqs, testutil.Contains, "sparc64/distro-other/rel-other/kver-other/base-other")

	c.Check(reqs, Not(testutil.Contains), "i386/ubuntu/rel-other/other-kver/base-other")
}
```

Reply:
Not strictly opposed to this, but the remaining tests are very small and easy to read, so I think I'd prefer not to add the magic.

Reply:
Yeah, I agree, as I said earlier I was getting carried away a bit and there it was definitely too far :)